File: Semestre_2/Vision_Computadora/Entregas/Matlab/T2_7_Especificacion_Histograma.m
close all;
clc;

[file,path] = uigetfile({'*.jpg';'*.png';'*.jpeg';'*.bmp';'*.gif'}, 'File Selector');
ruta = fullfile(path, file);

A = imread(ruta);
A = im2double(A);

figure;
subplot(231); imshow(A); title('Imagen A');
subplot(232); imhist(A); title('Histograma de A');

% Definir histograma de referencia. Se puede modificar por uniforme, gaussiano o exponencial. 
tipo_histograma = 'gaussiano'; 
switch lower(tipo_histograma)
    case 'uniforme'
        hist_ref = ones(256, 1) / 256; % Divide a la distribucion en nbins iguales
        
    case 'gaussiano'
        x = 0:255;  % Rango de x
        mu = 180;   % Media 
        sigma = 30; % Desviación estandar
        hist_ref = exp(-((x - mu).^2) / (2 * sigma^2))';
        hist_ref = hist_ref / sum(hist_ref);
        
    case 'exponencial'
        x = 0:255;
        lambda = 0.02;
        hist_ref = lambda * exp(-lambda * x)';
        hist_ref = hist_ref / sum(hist_ref);
end

% Calcular CDFs normalizando por el total de pixeles
cdf_A = cumsum(imhist(A)) / numel(A); 
cdf_ref = cumsum(hist_ref);

% Mapeo de intensidades usando interpolacion
R = zeros(size(A));
for i = 1:numel(A)
    [~, idx] = min(abs(cdf_ref - cdf_A(round(A(i)*255)+1)));
    R(i) = (idx-1)/255;
end

% Visualización de R
subplot(233); bar(0:255, hist_ref, 'FaceColor', [0.5 0.5 0.5]);
title('Histograma de referencia'); xlim([0 255]);
subplot(234); imshow(R); title('A transformada');
subplot(235); imhist(R); title('Histograma transformado');

File: Semestre_2/Vision_Computadora/Entregas/Python/T2_7_Especificacion_Histograma.py
import matplotlib.pyplot as plt
import numpy as np
from skimage.io import imread
from tkinter import Tk, filedialog

Tk().withdraw()
file_path = filedialog.askopenfilename(
    initialdir="C:\\Users\\death\\Documents\\Maestria\\MCIA\\Semestre_2\\Vision_Computadora\\img",
    filetypes=[("Images", "*.jpg;*.png;*.jpeg;*.bmp;*.gif")]
)

A = imread(file_path)

if A.ndim == 3 and A.shape[-1] == 4:
    A = A[..., :3]

A = A.astype(float) / 255.0

plt.figure()
plt.subplot(2,3,1); plt.imshow(A); plt.title('Imagen A'); plt.axis('off')

hist_A, _ = np.histogram(A.ravel(), bins=256, range=(0,1))
plt.subplot(2,3,2); plt.bar(np.arange(256), hist_A); plt.title('Histograma de A')

tipo_histograma = 'gaussiano'

if tipo_histograma.lower() == 'uniforme':
    hist_ref = np.ones(256) / 256.0
elif tipo_histograma.lower() == 'gaussiano':
    x = np.arange(256)
    mu = 180
    sigma = 30
    hist_ref = np.exp(-((x - mu)**2) / (2 * sigma**2))
    hist_ref = hist_ref / hist_ref.sum()
elif tipo_histograma.lower() == 'exponencial':
    x = np.arange(256)
    lam = 0.02
    hist_ref = lam * np.exp(-lam * x)
    hist_ref = hist_ref / hist_ref.sum()
else:
    hist_ref = np.ones(256) / 256.0

cdf_A = np.cumsum(hist_A) / hist_A.sum()
cdf_ref = np.cumsum(hist_ref)

T = np.zeros(256)
for g in range(256):
    idx = np.argmin(np.abs(cdf_ref - cdf_A[g]))
    T[g] = idx / 255.0

idx_img = np.round(A * 255).astype(int)
idx_img = np.clip(idx_img, 0, 255)
R = T[idx_img]

plt.subplot(2,3,3); plt.bar(np.arange(256), hist_ref); plt.title('Histograma de referencia'); plt.xlim(0,255)
plt.subplot(2,3,4); plt.imshow(R); plt.title('A transformada'); plt.axis('off')

hist_R, _ = np.histogram(R.ravel(), bins=256, range=(0,1))
plt.subplot(2,3,5); plt.bar(np.arange(256), hist_R); plt.title('Histograma transformado')

plt.tight_layout()
plt.show()
